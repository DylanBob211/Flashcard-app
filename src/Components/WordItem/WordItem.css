/* WordItem */
.wordItem_container{
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: var(--third-shade-secondary);
    cursor: pointer;
    padding: 2em 0;
    border-radius: 3px;
    margin-bottom: .5em;
}

.wordItem_name {
    margin-left: 1em;
    list-style-type: none;
    font-weight: 500;
    font-size: 1.2rem;
    text-shadow: -1px 1px 1px rgba(0,0,0,0.2);
    text-transform: capitalize;
    
}

.wordItem_remove{
    position: relative;
    background-color: #ff1818;
    margin-right: 1em;
    height: 2em;
    min-width: 2em;
    border-radius: 50%;
    cursor: pointer;
    transition: all .3s;
}
.wordItem_remove:hover{
    transform: scale(1.3)
}
.wordItem_remove::after{
    content: '';
    position: absolute;
    background-color: #fff;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 1;
    height: .25em;
    width: .7em;
    border-radius: 3px;
}



/* WordPreview */

.wordPreview_container {
    display: flex;
    flex-direction: column;
    text-align: center;
    justify-content: space-between;
    align-items: center;
    height: 50%;
    width: 20%;
    border-radius: 5px;
    background: #000;
    padding: .7em;
    z-index: 3;
    position: absolute;
}

.wordPreview_title{
    color: #fff;
    text-transform: capitalize;
}

.wordPreview_img {
    width: 90%;
    height: 90%;
}

.wordPreview_imgContainer{
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-auto-flow: dense;
    border: 2px solid #fff;
    padding: .3em;
}

.wordPreview_img--small{
    width: 50px;
    height: 50px;
}