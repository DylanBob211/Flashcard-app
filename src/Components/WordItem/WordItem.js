import React, { useState, useEffect } from 'react'
import './WordItem.css'

const WordItem = ({name, picUrls, deleteWord, id}) => {
    
    const [isMouseOver, setMouseOver] = useState(false)
    const [mousePosition, setMousePosition] = useState({mouseX : 0, mouseY: 0})

    const handleMouseOver = e => {
        setMouseOver(!isMouseOver)
    }
    const handleMouseMove = e => {
        setMousePosition({mouseX : e.clientX, mouseY: e.clientY })
    }

    return (
        <div onMouseOver={e => handleMouseOver(e)} onMouseOut={e => handleMouseOver(e)} onMouseMove={e => handleMouseMove(e)}>
            <WordPreview name={ name } picUrls={picUrls} isMouseOver={ isMouseOver } mousePosition={ mousePosition }/>
            <div className="wordItem_container">
                <li className="wordItem_name">{name}</li>
                <button className="wordItem_remove" onClick={e => deleteWord(name, id)}>-</button>        
            </div>
        </div>
    )
}

export default WordItem

const WordPreview = ({ picUrls, isMouseOver, name }) => {

    const dinamicStyle = {

        transition: isMouseOver ? '.5s all 1s' : '.3s all',
        transform: isMouseOver ? 'scaleZ(1) translate(10%, -101%)' : 'scaleZ(0) translate(10%, -101%)',
        transformOrigin: 'top',
        opacity: isMouseOver ? '0.9' : '0.2'
    }

    const renderImg = () => {
        if(picUrls === "No Img Available" || picUrls === ""){
            return <img className="wordPreview_img" src={ require('../../Assets/imgs/no_img.svg') } alt="no Word Preview"/>
        } else {
            return (
                <div className="wordPreview_imgContainer">
                    {picUrls.map((url, ind) => (<img key={ind} className="wordPreview_img--small" src={url} />))}
                </div>
            )
        }
    }

    return (
    <div className="wordPreview_container" style={dinamicStyle}>
        <h3 className="wordPreview_title">{ name }</h3>
        {renderImg()}    
    </div>
    )
}

