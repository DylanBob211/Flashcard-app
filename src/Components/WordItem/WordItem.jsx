import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import './WordItem.css';
import FlashcardPreview from './FlashcardPreview/FlashcardPreview';
import RemoveIcon from '../Icons/RemoveIcon';

const WordItem = ({
  wordItem, deleteWord, wordId, openFlashcard,
}) => {
  const [isMouseOver, setMouseOver] = useState(false);
  const handleDeleteClick = (e) => {
    e.stopPropagation();
    deleteWord(wordItem.word, wordId);
  };

  let animID;

  const handleMouseOver = () => {
    animID = setTimeout(() => setMouseOver(true), 1000);
  };

  const handleMouseOut = () => {
    if (animID) {
      clearTimeout(animID);
    }
    setMouseOver(false);
  };
  return (
    <div
      role="button"
      tabIndex={0}
      onMouseOver={handleMouseOver}
      onFocus={handleMouseOver}
      onMouseOut={handleMouseOut}
      onBlur={handleMouseOut}
      onClick={() => openFlashcard(wordItem)}
      onKeyPress={() => openFlashcard(wordItem)}
    >
      {isMouseOver && (
        <FlashcardPreview
          name={wordItem.word}
          imgUrls={wordItem.url}
          toggleVisibility={isMouseOver}
        />
      )}
      <div className="wordItem_container">
        <li className="wordItem_name">{wordItem.word}</li>
        <RemoveIcon
          className="wordItem_remove"
          disabled={!deleteWord}
          onClick={e => handleDeleteClick(e)}
        />
      </div>
    </div>
  );
};

WordItem.propTypes = {
  wordItem: PropTypes.shape({
    word: PropTypes.string.isRequired,
    url: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
  deleteWord: PropTypes.func,
  wordId: PropTypes.number.isRequired,
  openFlashcard: PropTypes.func,
};

WordItem.defaultProps = {
  deleteWord: null,
  openFlashcard: null,
};


export default WordItem;
