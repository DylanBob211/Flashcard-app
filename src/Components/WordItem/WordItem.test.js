import React from 'react';
import { mount, shallow } from 'enzyme';
import WordItem from './WordItem';
// What does it render?

// Does my component render different things under different circumstances?

// When i pass a function as a prop, what does my component use it for?
// Does it call it, or just give it to another component
// If it calls it, what does it call it with?

// When the user interacts with my component, what happens?

describe('WordItem', () => {
  let mountedWordItem;

  const wordItem = (props, isShallow = false) => {
    const jsx = (<WordItem {...props} />);
    if (isShallow) return shallow(jsx);
    return mount(jsx);
  };

  const props = {
    wordItem: {
      word: 'hello',
      url: [],
    },
    wordId: 0,
  };

  beforeEach(() => {
    mountedWordItem = undefined;
  });

  it('matches to snapshot', () => {
    expect(wordItem(props, true)).toMatchSnapshot();
  });

  describe('always', () => {
    beforeEach(() => {
      mountedWordItem = wordItem(props);
    });

    it('renders `word` prop in a `li` item', () => {
      const { word } = props.wordItem;
      const textContainer = mountedWordItem.find('.wordItem_name');
      expect(textContainer.text()).toEqual(word);
    });

    it('renders a RemoveIcon button', () => {
      const removeIcon = mountedWordItem.find('RemoveIcon');
      expect(removeIcon.exists()).toEqual(true);
    });

    it('disables the RemoveIcon button', () => {
      const removeIcon = mountedWordItem.find('RemoveIcon');
      expect(removeIcon.props().disabled).toEqual(true);
    });

    it('does not render FlashcardPreview', () => {
      const flashcardPreview = mountedWordItem.find('FlashcardPreview');
      expect(flashcardPreview.exists()).toEqual(false);
    });

    describe('when mouse is over', () => {
      beforeEach(() => {
        mountedWordItem.simulate('mouseover');
      });

      it('renders FlashcardPreview after one second', (done) => {
        const flashcardPreview = mountedWordItem.find('FlashcardPreview');
        expect(flashcardPreview.exists()).toEqual(false);

        setTimeout(() => {
          mountedWordItem.update();
          const flashcardPreview = mountedWordItem.find('FlashcardPreview');
          expect(flashcardPreview.exists()).toEqual(true);
          done();
        }, 1001);
      });

      it('stops rendering FlashcardPreview when mouse is out', () => {
        mountedWordItem.simulate('mouseout');
        const flashcardPreview = mountedWordItem.find('FlashcardPreview');
        expect(flashcardPreview.exists()).toEqual(false);
      });
    });

    describe('when clicking RemoveIcon button', () => {
      beforeEach(() => {
        props.deleteWord = jest.fn();
        props.openFlashcard = jest.fn();
        mountedWordItem = wordItem(props);
        const btn = mountedWordItem.find('RemoveIcon');
        btn.simulate('click');
      });

      it('calls deleteWord', () => {
        const { deleteWord, wordItem, wordId } = props;
        expect(deleteWord).toHaveBeenCalledWith(wordItem.word, wordId);
      });

      it('prevents bubbling', () => {
        expect(props.openFlashcard).not.toHaveBeenCalled();
      });
    });

    describe('when clicking the whole element', () => {
      beforeEach(() => {
        props.deleteWord = jest.fn();
        props.openFlashcard = jest.fn();
        mountedWordItem = wordItem(props);
        mountedWordItem.simulate('click');
      });

      it('calls openFlashcard', () => {
        const { wordItem, openFlashcard } = props;
        expect(openFlashcard).toHaveBeenCalledWith(wordItem);
      });
    });
  });
});
