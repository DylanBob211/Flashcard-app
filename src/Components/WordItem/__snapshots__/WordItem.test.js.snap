// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordItem matches to snapshot 1`] = `
<div
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="wordItem_container"
  >
    <li
      className="wordItem_name"
    >
      hello
    </li>
    <RemoveIcon
      className="wordItem_remove"
      disabled={true}
      onClick={[Function]}
    />
  </div>
</div>
`;
